azure-iot-device/tests/common/pipeline/pipeline_ops_test.py

```diff
@@ -0,0 +1,461 @@
+
+            cb1_mock.side_effect = cb1
+            cb2_mock.side_effect = cb2
+            cb3_mock.side_effect = cb3
+
+            # Attach callbacks to op
+            init_kwargs["callback"] = cb1_mock
+            op = cls_type(**init_kwargs)
+            op.add_callback(cb2_mock)
+            op.add_callback(cb3_mock)
+            assert len(op.callbacks) == 3
+            assert not op.completed
+
+            # Run the completion
+            op.complete(error=error)
+
+            assert op.completed
+            assert cb3_mock.call_count == 1
+            assert cb3_mock.call_args == mocker.call(op=op, error=error)
+            assert cb2_mock.call_count == 1
+            assert cb2_mock.call_args == mocker.call(op=op, error=error)
+            assert cb1_mock.call_count == 1
+            assert cb1_mock.call_args == mocker.call(op=op, error=error)
+
+            # Because exceptions raised in callbacks are caught and sent to the background exception handler,
+            # the assertions in the above callbacks won't be able to directly raise AssertionErrors that will
+            # allow for testing normally. Instead we should check the background_exception_handler to see if
+            # any of the assertions raised errors and sent them there.
+            assert handle_exceptions.handle_background_exception.call_count == 0
+
+        @pytest.mark.it(
+            "Handles any Exceptions raised during execution of a callback by sending them to the background exception handler, and continuing on with completion"
+        )
+        def test_callback_raises_error(
+            self, mocker, arbitrary_exception, cls_type, init_kwargs, error
+        ):
+            mocker.spy(handle_exceptions, "handle_background_exception")
+
+            # Set up callback mocks
+            cb1_mock = mocker.MagicMock()
+            cb2_mock = mocker.MagicMock(side_effect=arbitrary_exception)
+            cb3_mock = mocker.MagicMock()
+
+            # Attach callbacks to op
+            init_kwargs["callback"] = cb1_mock
+            op = cls_type(**init_kwargs)
+            op.add_callback(cb2_mock)
+            op.add_callback(cb3_mock)
+            assert len(op.callbacks) == 3
+            assert not op.completed
+
+            # Run the completion
+            op.complete(error=error)
+
+            # Op was completed, and all callbacks triggered despite the callback raising an exception
+            assert op.completed
+            assert len(op.callbacks) == 0
+
+            # The exception raised by the callback was passed to the background exception handler
+            assert handle_exceptions.handle_background_exception.call_count == 1
+            assert handle_exceptions.handle_background_exception.call_args == mocker.call(
+                arbitrary_exception
+            )
+
+        @pytest.mark.it(
+            "Allows any BaseExceptions raised during execution of a callback to propagate"
```

Comment:
>propagate [](start = 80, length = 9)

Do you want to add asserts that verify that cb3_mock was called but cb1_mock was not called?

Reply:
Good call.

---
In reply to: [346506390](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r346506390) [](ancestors = 346506390)